lib/Conch/Controller/User.pm

```diff
@@ -442,6 +442,100 @@ sub deactivate ($c) {
 	return $c->status(204);
 }
 
+=head2 get_tokens
+
+Get a list of user tokens. By default only returns active tokens; pass C<?with_expired=1>
+to include expired tokens as well.
+
+Response uses the UserTokens json schema.
+
+=cut
+
+sub get_tokens ($c) {
+    my $rs = $c->db_user_session_tokens->order_by('name');
+    $rs = $rs->unexpired if not $c->req->query_params->param('with_expired') // 0;
+
+    return $c->status(200, [ $rs->all ]);
+}
+
+=head2 create_token
+
+Create a new token, creating a JWT from it.  Response uses the NewUserToken json schema.
+
+=cut
+
+sub create_token ($c) {
+    my $input = $c->validate_input('NewUserToken');
+    return if not $input;
+
+    return $c->status(400, { error => 'name "'.$input->{name}.'" is already in use' })
+        if $c->db_user_session_tokens
+            ->search({ user_id => $c->stash('user_id'), name => $input->{name} })->exists;
+
+    # default expiration: 5 years
+    my $expires_abs = time + (($c->config('jwt') || {})->{custom_token_expiry} // 86400*365*5);
+
+    # TODO: ew ew ew, some duplication with Conch::Controller::Login::_create_jwt.
+    my ($new_db_row, $token) = $c->db_user_session_tokens->generate_for_user(
+        $c->stash('user_id'), $expires_abs, $input->{name});
+
+    my $jwt = Mojo::JWT->new(
+        claims => { uid => $c->stash('user_id'), jti => $token },
+        secret => $c->config('secrets')->[0],
+        expires => $expires_abs,
+    )->encode;
+
+    return $c->status(201, {
+        token => $jwt,
+        $new_db_row->TO_JSON->%*,
+    });
+}
+
+=head2 find_token
+
+Chainable action that takes the 'token_name' provided in the path and looks it up in the
+database, stashing a resultset to access it as 'token_rs'.
+
+Only the current user may access the token.
+
+=cut
+
+sub find_token ($c) {
+    my $token_rs = $c->db_user_session_tokens->unexpired
+        ->search({ user_id => $c->stash('user_id'), name => $c->stash('token_name') });
+
```

Comment:
Does this mean we can list expired token rows but we can't look them up individually?

Reply:
yes. I can add `?with_expired=1` support to the individual GET request as well.

Note that in practice this will be pretty pointless though, as (as noted previously) we automatically delete all expired tokens when passing through the login code.

Reply:
Yeah. I don't like having one endpoint say "here's this data" and another endpoint say "what data? 404!". You're right that we can't _avoid_ that if someone issues fetches the list of expired tokens and then we delete them before they can fetch a single token ... but we can at least try to minimize it.

Reply:
Let's look at it from a different direction. Why are we allowing users to access expired tokens at all, particularly if at some random point the code is going to delete them out from under us?